Jenkinsfile

```diff
@@ -501,37 +501,43 @@ pipeline {
                         }
                     }
                     steps {
+                        // cycle through debug, release, and dev builds
                         sconsBuild clean: "_build.external${arch}",
-                                   failure_artifacts: 'config.log-centos7-gcc'
+                                   failure_artifacts: 'config.log-centos7-gcc-debug',
+                                   BUILD_TYPE: 'debug'
+                        sconsBuild skip_clean: '1', failure_artifacts: 'config.log-centos7-gcc-release',
+                                   BUILD_TYPE: 'release'
+                        sconsBuild skip_clean: '1', failure_artifacts: 'config.log-centos7-gcc-dev',
+                                   BUILD_TYPE: 'dev'
```

Comment:
Wondering about adding the Coverity build here, and the future Bullseye build.
Both of them only need to build DAOS.
Bullseye produces an additional set of artifacts.

Reply:
For Bullseye, the build will go to build/*/covc

Not sure about Coverity.  How does coverity work?

Reply:
If we add more builds, we should probably spread the load to other platforms

Reply:
These serialised builds are adding 26 minutes to the build stage for EL7.  What problem(s) are the additional 2 builds solving and are they worth adding a half hour to everyone's PRs?

How can we still satisfy Quick-build's goal of doing the most minimum/quickest build possible, to make iterating in CI more efficient?

Reply:
For an upcoming change, we are going to need all the build steps in a stage to be almost identical.
So this is all speculation for a different PR.

Yes, bullseye just needs the binary with license key to be built into the dockerfile.
For a common Jenkinsfile to be used outside of our build system, we would need to be able to detect if a bullseye build is allowed in the sconsBuild step, and ignore the request for a 'cov' build in that case.

I am now wondering about changing BUILD_TYPE to be build_types that takes an array of steps, ['debug'. 'release', 'dev']

Coverity build is same as BUILD_TYPE 'dev' with the coverity parameter filled in.

So for the above, I would be adding a 'coverity' value the build_types array, and change sconsBuild to ignore the coverity parameter both for when we can not download the coverity software, or when build_types parameter is present, but does not include it.

The failure_artifacts can default to a generated string.

Reply:
I don't want to see additional build-time for PRs.

What I was thinking was that we could build all of debug, release, and dev but that we could also spread them across the different distros/compiler versions, in effect a sparse matrix.

Reply:
We can not currently add any more build stages to DAOS due to jenkinsfile size limitations.

Jenkins Declarative Matrix pipeline builds are not reporting properly in the web UIs for sparse matrix builds, and would still require all build steps in parallel to run the exact same stages and steps, which will require a refactoring of the Jenkinsfile.

The alternative that we are going to have to go with is a similar refactoring, so we have build stages with as few steps as possible, which means both moving muliple steps into pipeline-lib and moving scripts into DAOS.  There is another ticket for that and some test PRs in progress.

Once we get past that issue, we can use the quickbuild option currently for all builds that do not need to produce artifacts that need to used in tests.

Reply:
26 minutes?  It’s only two extra builds.  I thought it was closer to 12 since they don’t rebuild from scratch.

I’m fine with either way

Reply:
On a side note, are the build directories on a local disk?  If not, we could probably use BUILD_ROOT to speed things up a bit

Reply:
case in point, on my box, it takes 1 minute, 20 seconds to build just daos with -j 4.   With -j 24 it takes 33 seconds.   Granted 24 is probably much for the docker files but if we are on nfs, it is likely slowing us down quite a bit.

Reply:
I assume the 24 minutes includes the full dependency list?  If I understand Jeffs change correctly it'll just rebuild daos for the different build types.

Reply:
> 26 minutes?

[Yes](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2601/10/pipeline/92).  As you can see, the full component build is 21m35s, the _release_ build is 12m55s and the _dev_ build is 13m5s.

> It’s only two extra builds. I thought it was closer to 12 since they don’t rebuild from scratch.

~12m **each**.

> On a side note, are the build directories on a local disk?

Yes.

> case in point, on my box, it takes 1 minute, 20 seconds to build just daos with -j 4.

We don't provide any multi-CPU flags (`-j ...`) in CI.

> With -j 24 it takes 33 seconds. Granted 24 is probably much for the docker files

Maybe.  Maybe not.  One could be of the opinion that if you give every build the ability to use as many CPUs as are available, one will get the fastest builds at times of low use/contention and in times of high use/contention, the kernel will portion out the available CPU to all users equally/accordingly.

> but if we are on nfs,

We are not.

> Without -j option, I get 4 minutes, 44 seconds on my box.

What kind of disk do you have?  I'm not sure I'm quite understanding the composition of the disk under the docker containers, but it seems to be made of both an INTEL SSDPEDMD400G4 NVME disk and an INTEL SSDSC2BA200G4 SSD.  Not sure if those are being used in a simple linear concatenation or it's tiered storage, etc.

> I assume the 24 minutes

I assume you are referring to the additional _26_ minutes, so my further comments are based on that assumption.

> includes the full dependency list?

From the looks of at least [one of the extra builds](https://build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/daos/branches/PR-2601/runs/10/nodes/92/steps/290/log/?start=0), no, that is just DAOS.  I would assume the other build is the same.

> If I understand Jeffs change correctly it'll just rebuild daos for the different build types.

Correct, and that's what it seems to be doing, with only a single CPU core though, no `-j ...`.

[Enabling parallel jobs in scons](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-2626/1/pipeline/59) (and it's children) brings build times down pretty significantly.

The non-RPM CentOS 7 build goes down form ~21.5m to 6m, with the release and dev builds going down to 2m5s and 2m6s from the ~13m for non-parallel builds.

The RPM builds go down to 4m20s and 3m31s from 11m46 and 9m31 for EL7 and Leap15, respectively.

Now, that is on what is probably a pretty quiet system, at a quite time of day, so those probably got all 72 CPUs.  But maybe given the short durations, it's what most builds will experience most of the time.

It certainly seems worthwhile to enable this generally, I think.  I'll push another patch to my PR to enable the parallel-building on the rest of the stages.

Reply:
That's on a normal spinning disk.    My box has an SSD, I should try using that as BUILD_ROOT

Reply:
Parallel builds work for all of our PR combinations except [_Build on Leap 15 with Intel-C and TARGET_PREFIX_](https://build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/daos/branches/PR-2626/runs/4/nodes/68/steps/178/log/?start=0).

Seems like that would be a bug of some sort then.

Reply:
> I'm seeing a full build take 2m24s

Full build including all components?

Reply:
Actually, in looking at my configuration, I use /home for build but in gparted, it shows fedora-home is an lvm and it appears to use the ssd as well as the spinning disk.

Reply:
> > I'm seeing a full build take 2m24s
> 
> Full build including all components? Just a daos build, no components on a single core on my _Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz_ with an NVMe drive takes 3m8s.
> 
> Are you sure you are not doing parallel builds and just not realising it, or using ccache or some such?

"scons" and then "scons install" are the two commands I'm timing, dependencies are already built.  No ccache or similar like that.  A "debug" build using clang on CentOS 7.7.  It's on xfs over LVM.  The source tree is imported from OS X to Linux using parallels, PREFIX, PREBUILT_PREFIX and BUILD_ROOT are all set to different directories in /home.  2 cores, and 8Gb ram configured.

Reply:
The docker builders are building on spinning rust.  No NVMe or SSD drives there.